cli/src/pcluster/validators/cluster_validators.py

```diff
@@ -417,42 +417,78 @@ def _validate(
+    Each frozen set contains sg combination of a queue.
     :param protocol: the IP protocol to be checked.
     :return: True if both in and out access are allowed
     :raise: ClientError if a given security group doesn't exist
     """
     in_access = False
     out_access = False
+    src_ip_ranges = []
+    dst_ip_ranges = []
+    src_security_groups = set()
+    dst_security_groups = set()
 
     for sec_group in AWSApi.instance().ec2.describe_security_groups(security_groups_ids):
         # Check all inbound rules
         for rule in sec_group.get("IpPermissions"):
-            if _check_sg_rules_for_port(rule, port, protocol):
-                if is_cidr_optional or rule.get("IpRanges") or rule.get("PrefixListIds"):
-                    in_access = True
-                    break
+            if in_access:
+                break
+            if _is_port_allowed_by_sg_rule(rule, port, protocol):
+                in_access = _populate_allowed_src_or_dst(rule, src_ip_ranges, src_security_groups)
 
         # Check all outbound rules
         for rule in sec_group.get("IpPermissionsEgress"):
-            if _check_sg_rules_for_port(rule, port, protocol):
-                if is_cidr_optional or rule.get("IpRanges") or rule.get("PrefixListIds"):
-                    out_access = True
-                    break
+            if out_access:
+                break
+            if _is_port_allowed_by_sg_rule(rule, port, protocol):
+                out_access = _populate_allowed_src_or_dst(rule, dst_ip_ranges, dst_security_groups)
 
         if in_access and out_access:
             return True
-
+    # If in_access or out_access is still False, check allowed ip ranges and security groups.
+    # The in_access or out_access could only have been true, if previous logics had found prefix list in SG rules.
+    # Rules of ip ranges have to be checked at the end because the union of all ip ranges may cover the subnets,
+    # even when individual ip ranges do not cover the subnets. The same reason applies to allowed security groups.
+    in_access = in_access or _are_ip_ranges_and_sg_accessible(
+        security_groups_by_nodes, src_ip_ranges, src_security_groups, subnets
+    )
+    out_access = out_access or _are_ip_ranges_and_sg_accessible(
+        security_groups_by_nodes, dst_ip_ranges, dst_security_groups, subnets
+    )
+    return in_access and out_access
+
+
+def _are_ip_ranges_and_sg_accessible(security_groups_by_nodes, allowed_ip_ranges, allowed_security_groups, subnets):
+    # For all cluster nodes, at least one of the security groups attached need to be in the UserIdGroupPairs.
+    return all(
+        node_security_groups & allowed_security_groups for node_security_groups in security_groups_by_nodes
+    ) or _are_subnets_covered_by_cidrs(allowed_ip_ranges, subnets)
+
+
+def _populate_allowed_src_or_dst(rule, ip_ranges, allowed_security_groups):
```

Comment:
a function called `_populate_...` is unexpected to return a boolean value.
Is returning a boolean value required at all?

Reply:
The minor inconsistency comes from the special case of `PrefixListIds`. I would keep the name because the majority of the function is doing population. The boolean return is to handle a special case

Reply:
May you please add a brief comment on the function to explain the behaviour, especially about the meaning of the boolean value?

Reply:
Done